src/node_errors.cc

```diff
@@ -364,57 +364,61 @@ static void ReportFatalException(Environment* env,
       // Not an error object. Just print as-is.
       String::Utf8Value message(env->isolate(), error);
 
-      PrintErrorString("%s\n",
-                       *message ? *message : "<toString() threw exception>");
+      PrintToStderr(env->event_loop(),
+                    "%s\n",
+                    *message ? *message : "<toString() threw exception>");
     } else {
       node::Utf8Value name_string(env->isolate(), name.ToLocalChecked());
       node::Utf8Value message_string(env->isolate(), message.ToLocalChecked());
 
       if (arrow.IsEmpty() || !arrow->IsString() || decorated) {
-        PrintErrorString("%s: %s\n", *name_string, *message_string);
+        PrintToStderr(
+            env->event_loop(), "%s: %s\n", *name_string, *message_string);
       } else {
         node::Utf8Value arrow_string(env->isolate(), arrow);
-        PrintErrorString(
-            "%s\n%s: %s\n", *arrow_string, *name_string, *message_string);
+        PrintToStderr(env->event_loop(),
+                      "%s\n%s: %s\n",
+                      *arrow_string,
+                      *name_string,
+                      *message_string);
       }
     }
   }
 
   fflush(stderr);
 }
 
-void PrintErrorString(const char* format, ...) {
+int PrintToStderr(uv_loop_t* loop, const char* format, ...) {
   va_list ap;
   va_start(ap, format);
 #ifdef _WIN32
-  HANDLE stderr_handle = GetStdHandle(STD_ERROR_HANDLE);
-
+  int stderr_fileno = _fileno(stderr);
   // Check if stderr is something other than a tty/console
-  if (stderr_handle == INVALID_HANDLE_VALUE || stderr_handle == nullptr ||
-      uv_guess_handle(_fileno(stderr)) != UV_TTY) {
+  if (uv_guess_handle(stderr_fileno) != UV_TTY) {
     vfprintf(stderr, format, ap);
     va_end(ap);
-    return;
+    return 0;
   }
 
   // Fill in any placeholders
   int n = _vscprintf(format, ap);
-  std::vector<char> out(n + 1);
+  std::vector<char> out(n + 1, 0);
   vsprintf(out.data(), format, ap);
 
-  // Get required wide buffer size
-  n = MultiByteToWideChar(CP_UTF8, 0, out.data(), -1, nullptr, 0);
-
-  std::vector<wchar_t> wbuf(n);
-  MultiByteToWideChar(CP_UTF8, 0, out.data(), -1, wbuf.data(), n);
-
-  // Don't include the null character in the output
-  CHECK_GT(n, 0);
-  WriteConsoleW(stderr_handle, wbuf.data(), n - 1, nullptr, nullptr);
+  uv_buf_t wrbuf = uv_buf_init(out.data(), n);
+  uv_tty_t handle;
+  uv_tty_init(loop, &handle, stderr_fileno, false);
```

Comment:
Do we have a test case that verifies that this works even when `process.stderr` has also been accessed? IIRC libuv doesn’t allow multiple handles per fd as part of its API contract…

Reply:
> Do we have a test case that verifies that this works even when process.stderr has also been accessed?

I guess the easiest way is to use `process._rawDebug` in the test? I tried it locally with a test in `pseudo-tty` and it worked.

> IIRC libuv doesn’t allow multiple handles per fd as part of its API contract…

I think that only applies to `uv_poll_t`? https://github.com/joyent/libuv/issues/680 We are using `uv_try_write` here so it's not going to be queued, it's also only used on Windows.

Maybe we could also consider exposing the color code translation, or writing into a fd with that directly from libuv instead of going with this. cc @bnoordhuis

Reply:
uh, right, `pseudo-tty` tests are skipped on Windows, I tried running the test directly from console and it indeed asserted

```
Assertion failed: 0, file c:\users\joyee\projects\node\deps\uv\src\win\handle-inl.h, line 160
```

Reply:
Just to confirm: libuv currently does not allow more than one handle per file descriptor. Relevant FR: https://github.com/libuv/libuv/issues/1172

> Maybe we could also consider exposing the color code translation

Maybe. It's not abstracted in a way that allows easy exposing right now.

> or writing into a fd with that directly from libuv 

I'm not sure what you mean.